File: 0826-most-profit-assigning-work/0826-most-profit-assigning-work.py
class Solution:
    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        # Find maximum ability in the worker array
        max_ability = max(worker)
        jobs = [0] * (max_ability + 1)
        
        for i in range(len(difficulty)):
            if difficulty[i] <= max_ability:
                jobs[difficulty[i]] = max(jobs[difficulty[i]], profit[i])
        
        # Take maxima of prefixes
        for i in range(1, max_ability + 1):
            jobs[i] = max(jobs[i], jobs[i - 1])
            
        net_profit = 0
        for ability in worker:
            net_profit += jobs[ability]
            
        return net_profit
            

File: 0826-most-profit-assigning-work/0826-most-profit-assigning-work.cpp
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        vector<pair<int, int>> jobProfile{{0, 0}};
        
        for (int i = 0; i < difficulty.size(); ++i)
            jobProfile.push_back({profit[i], difficulty[i]});
        
        // sort in decreasing order of profit
        sort(jobProfile.begin(), jobProfile.end(), [](pair<int, int>& p1, pair<int, int>& p2){return p1.first > p2.first;});
        
        for (int i = 1; i < jobProfile.size(); ++i)
            jobProfile[i] = pair<int, int>{jobProfile[i].first, min(jobProfile[i - 1].second, jobProfile[i].second)};
        
        int netProfit = 0;
        for (int ability : worker){
            int l = 0, r = jobProfile.size() - 1, jobProfit = 0;
            
            // Maximize profit using binary search
            while (l <= r){
                int mid = l + ((r - l) >> 1);
                
                if (jobProfile[mid].second <= ability){
                    r = mid - 1;
                    jobProfit = max(jobProfit, jobProfile[mid].first);
                }
                else l = mid + 1;
            }

            // Add profit of each worker to total profit.
            netProfit += jobProfit;
        }
        
        return netProfit;
    }
};